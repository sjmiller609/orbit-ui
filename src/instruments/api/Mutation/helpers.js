const has = (str, find) => ~str.toLowerCase().indexOf(find.toLowerCase())

export const formErrors = ({ error, errors }) => {
  if (!error) return

  let err = JSON.stringify(error).toLowerCase()
  return errors.find(e => {
    if (e.key && has(err, e.key)) return e
    else if (e.allKeys && e.allKeys.every(k => has(err, k))) return e
    else if (e.someKeys && e.someKeys.some(k => has(err, k))) return e
  })
}
