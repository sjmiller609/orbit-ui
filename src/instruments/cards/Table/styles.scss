@import '../../../styles/mixins.scss';

.search {
  padding-right: 3rem;
}

@media screen and (max-width: 680px) {
  .header {
    flex-wrap: wrap;
  }
}

.tableRow {
  display: block;
  padding: 1.5rem 0.5rem 1.5rem 1rem;

  &:nth-child(even) {
    background-color: $lightGray;
  }
}

.arrow {
  height: 1.1rem;
  opacity: 0.4;
}

.fields > div,
.arrowCol {
  padding: 0.5rem 1rem;
}
