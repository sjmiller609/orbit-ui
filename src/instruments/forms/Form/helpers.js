// faltten and stringify nested keys
export const unpack = (obj, ns, hash) => {
  if (!obj) return
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keypath = ns ? `${ns}.${key}` : key
    if (value && typeof value === 'object') {
      return unpack(value, keypath, acc)
    }
    acc[keypath] = value
    return acc
  }, hash || {})
}

const roll = (keys, obj, value) => {
  const key = keys[0]
  // case 1 - not obj
  if (keys.length === 1) {
    obj[key] = value
    return obj
  }
  // case 2 - doesn't exist
  if (!obj[key]) {
    obj[key] = isNaN(keys[1]) ? {} : []
  }
  // case 3 - exists
  obj[key] = roll(
    keys.slice(1),
    Array.isArray(obj[key]) ? [...obj[key]] : { ...obj[key] },
    value
  )

  return obj
}

// convert back into obj
export const pack = (obj, type = {}) => {
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keys = key.split('.')

    return roll(keys, acc, value)
  }, type)
}

export const packChild = ({ name, obj }) => {
  const obj2 = {}
  let isArray
  Object.entries(obj).forEach(k => {
    if (name.length < k[0].length && k[0].indexOf(name) === 0) {
      const n = k[0].slice(name.length + 1)
      // must be true for all matching
      isArray = !isNaN(n.slice(0, n.indexOf('.')))
      obj2[n] = k[1]
    }
  })
  if (Object.keys(obj2).length) return pack(obj2, isArray ? [] : {})
  return null
}

export const removeChild = ({ name, obj = {} }) => {
  const obj2 = {}
  Object.entries(obj).forEach(k => {
    if (name.length >= k[0].length || k[0].indexOf(name) !== 0) {
      obj2[k[0]] = k[1]
    }
  })
  return obj2
}
