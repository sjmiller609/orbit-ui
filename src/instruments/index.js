// NOTE: Any instruments importing other instruments must come AFTER

export { default as Row } from './flex/Row'
export { default as Box } from './flex/Box'

export { default as Card } from './cards/Card'
export { default as Table } from './cards/Table'
export { default as TableRow } from './cards/Table/TableRow'
export { default as CardForm } from './cards/CardForm'

export { default as Module } from './layouts/Module'
export { default as Header } from './layouts/Header'

export { default as Loading } from './animations/Loading'
export { default as LoadingDots } from './animations/LoadingDots'

export { default as Search } from './forms/Search'
export { default as Form } from './forms/Form'
export { default as TextField } from './forms/TextField'
export { default as Password } from './forms/Password'
export { default as Checkbox } from './forms/Checkbox'

export { default as Link } from './links/Link'
export { default as Button } from './links/Button'

export { default as Icon } from './icons/Icon'
export { default as Logo } from './icons/Logo'

export { default as P } from './text/P'
export { default as Mini } from './text/Mini'
export { default as ShowDate } from './text/ShowDate'
export { default as B } from './text/B'

export { default as H1 } from './text/H1'
export { default as H2 } from './text/H2'
export { default as H3 } from './text/H3'
export { default as H4 } from './text/H4'
export { default as H5 } from './text/H5'

export { default as Menu } from './menus/Menu'
export { default as Level1 } from './menus/Level1'

export { default as Snackbar } from './ui/Snackbar'

export { default as ContextUI } from './context/UI/Get'
export { default as ProviderUI } from './context/UI'

export { default as Query } from './api/Query'
export { default as Mutation } from './api/Mutation'
