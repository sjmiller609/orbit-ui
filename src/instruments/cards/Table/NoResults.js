'use strict'
import React from 'react'

import { Box, H5, Icon } from '../../../instruments'
import s from './styles.scss'

const NoResults = () => {
  return (
    <Box className={s.noResults}>
      <Icon icon="alien_ship" />
      <H5>Nothing found.</H5>
    </Box>
  )
}

export default NoResults
