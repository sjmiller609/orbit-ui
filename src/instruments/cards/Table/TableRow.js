'use strict'
import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import s from './styles.scss'
import { Row, Box, Link, Icon } from 'instruments'

const TableRow = ({ columns, to, className, wrap }) => {
  const row = (
    <Row justify="space-between">
      <Row justify="flex-start" className={s.fields} wrap={wrap} full>
        {Array.isArray(columns) ? columns.map(el => el) : columns}
      </Row>
      {to && (
        <Box auto className={s.arrowCol}>
          <Icon key="arrow" icon="arrow" className={s.arrow} />
        </Box>
      )}
    </Row>
  )
  if (!to) return row
  return (
    <Link to={to} className={classnames(s.tableRow, className)}>
      {row}
    </Link>
  )
}

TableRow.propTypes = {
  columns: PropTypes.oneOfType([PropTypes.array, PropTypes.element]),
  to: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  className: PropTypes.string,
  wrap: PropTypes.bool,
}

export default TableRow
