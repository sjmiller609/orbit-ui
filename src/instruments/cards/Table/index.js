'use strict'
import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { Card, Search } from '../../../instruments'
import s from './styles.scss'

const Table = ({ children, className }) => {
  return (
    <Card
      className={classnames(s.table, className)}
      header={
        <div className={s.header}>
          <Search />
        </div>
      }>
      <React.Fragment>
        {Array.isArray(children) ? children.map(el => el) : children}
      </React.Fragment>
    </Card>
  )
}

Table.propTypes = {
  children: PropTypes.oneOfType([PropTypes.array, PropTypes.element]),
  className: PropTypes.string,
}

export default Table
