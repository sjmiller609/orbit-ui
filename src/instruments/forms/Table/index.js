import PropTypes from 'prop-types'
import React from 'react'
import s from './styles.scss'
import classnames from 'classnames'
import { Field, TextField, Mini, TextButton } from 'instruments'
import { jsonEqual } from 'helpers/compare'

class Table extends React.Component {
  validate = this.validate.bind(this)
  remove = this.remove.bind(this)
  add = this.add.bind(this)
  renderRow = this.renderRow.bind(this)
  removeEmpties = this.removeEmpties.bind(this)

  fieldProps = this.fieldProps.bind(this)

  componentWillMount() {
    this.add()
  }

  componentDidMount() {
    const { value, registerOnSubmit, name } = this.props
    this.validate(value) // adds required fields to form
    if (registerOnSubmit)
      registerOnSubmit({ name, onSubmit: this.removeEmpties })
  }

  componentWillReceiveProps({ value }) {
    if (!jsonEqual(value, this.props.value)) {
      // run validation
      this.validate(value)
    }
  }

  removeEmpties(data) {
    if (!data) return
    return data.filter(d => !!d)
  }

  validate(value) {
    const { name, validate, updateErrors } = this.props
    let e
    if (validate) {
      // only validation is if passed in as prop
      e = validate(value)
      updateErrors(name, e)
    }
  }

  fieldProps(i) {
    const { className, ...props } = this.props.fieldProps
    return {
      ...props,
      ...this.props.formField(this.props.name + '.' + i),
      className: classnames(s.field, className),
      required: this.props.required,
      formField: this.props.formField,
    }
  }

  remove(i) {
    const { onChange, value } = this.props
    const v2 = Array.from(value)
    v2.splice(i, 1)
    onChange(null, v2)
  }

  add() {
    const { onChange, value } = this.props
    const v2 = Array.from(value)
    v2.push(null)
    onChange(null, v2)
  }

  renderRow(v, i) {
    const { Row, getRowProps, value } = this.props
    if (i === value.length - 1) return null
    return (
      <Row
        key={i}
        {...getRowProps(v)}
        className={s.tableRow}
        remove={() => this.remove(i)}
      />
    )
  }

  render() {
    const { value, id, title, className, Empty, FieldType } = this.props
    let count = value ? value.length - 1 : 0
    if (count === 1 && !value[0]) count = 0
    console.log(value)
    return (
      <div id={id} className={classnames(s.fields, className)}>
        {count > 0 ? (
          <React.Fragment>
            <div className={classnames(s.tableBorder, count > 1 && s.border)}>
              <div className={classnames(s.table, count === 1 && s.one)}>
                {value.map((v, i) => this.renderRow(v, i))}
              </div>
            </div>
            {title &&
              count > 1 && (
                <Mini>
                  {count} {title.toLowerCase()}
                  {count > 1 ? 's' : ''}
                </Mini>
              )}
          </React.Fragment>
        ) : (
          Empty && <Empty />
        )}

        <div className={s.field}>
          <FieldType {...this.fieldProps(value.length - 1)} />
          <TextButton className={s.add} onClick={this.add}>
            Add Another
          </TextButton>
        </div>
      </div>
    )
  }
}

Table.propTypes = {
  fieldProps: PropTypes.object,
  FieldType: PropTypes.func,
  formField: PropTypes.func,
  Row: PropTypes.func,
  Empty: PropTypes.func,
  getRowProps: PropTypes.func,
  value: PropTypes.array,
  name: PropTypes.string.isRequired,
  id: PropTypes.string,
  validate: PropTypes.func,
  updateErrors: PropTypes.func,
  onChange: PropTypes.func,
  required: PropTypes.bool,
  className: PropTypes.string,
  title: PropTypes.string,
  registerOnSubmit: PropTypes.func,
}

Table.defaultProps = {
  FieldType: TextField,
  fieldProps: {},
  value: [],
}

export default Field(Table)
