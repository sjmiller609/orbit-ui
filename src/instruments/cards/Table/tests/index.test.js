import React from 'react'
import { shallow } from 'enzyme'
import { mountWrap } from 'helpers/tests'
import { Link } from 'instruments'

import Table from '../index'
import TableRow from '../TableRow'
import NoResults from '../NoResults'

const remove = jest.fn()

const props = {
  children: ['<p>This is a test</p>'],
  className: 'testClassString',
  search: {
    call: jest.fn(),
    text: 'Test search',
    placeholder: 'Test search',
    delay: false,
  },
  button: {},
  Empty: jest.fn(() => React.Fragment),
  headerOptions: <Link>Test</Link>,
  nav: <Link>Test</Link>,
}

const rowProps = {
  columns: ['1', '2', '3'],
  to: '/',
  className: 'testClassString',
  wrap: false,
  remove,
}

const rowPropsElCol = {
  ...rowProps,
  columns: <p key="1">1</p>,
}

const rowPropsNoTo = {
  ...rowProps,
  to: null,
}

describe('Table', () => {
  let wrapper
  let tableRow
  let noResults

  beforeEach(() => {
    wrapper = mountWrap(<Table {...props} />)
    tableRow = shallow(<TableRow {...rowProps} />)
    noResults = mountWrap(<NoResults />)
  })

  it('should be defined', () => {
    expect(Table).toBeDefined()
    expect(TableRow).toBeDefined()
    expect(NoResults).toBeDefined()
  })

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
    expect(tableRow).toMatchSnapshot()
    expect(noResults).toMatchSnapshot()
  })

  it('should render elements for rows', () => {
    tableRow = shallow(<TableRow {...rowPropsElCol} />)
    expect(tableRow).toMatchSnapshot()
  })

  it('should render default to', () => {
    tableRow = shallow(<TableRow {...rowPropsNoTo} />)
    expect(tableRow).toMatchSnapshot()
  })

  it('should handle no button', () => {
    tableRow = shallow(<Table {...props} button={null} />)
    expect(tableRow).toMatchSnapshot()
  })
})
