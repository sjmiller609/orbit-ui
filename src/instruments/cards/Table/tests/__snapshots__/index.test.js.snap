// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should handle no button 1`] = `
<Card
  className="table one testClassString"
  header={
    <React.Fragment>
      <Link>
        Test
      </Link>
      <Row
        className="header"
        justify="space-between"
      >
        <Search
          className="search"
          dark={false}
          noDelay={true}
          placeholder="Test search"
          search={[MockFunction]}
          text="Test search"
        />
        <Link>
          Test
        </Link>
      </Row>
    </React.Fragment>
  }
>
  &lt;p&gt;This is a test&lt;/p&gt;
</Card>
`;

exports[`Table should render correctly 1`] = `
<Table
  Empty={[MockFunction]}
  button={Object {}}
  className="testClassString"
  headerOptions={
    <Link>
      Test
    </Link>
  }
  nav={
    <Link>
      Test
    </Link>
  }
  search={
    Object {
      "call": [MockFunction],
      "delay": false,
      "placeholder": "Test search",
      "text": "Test search",
    }
  }
>
  <Card
    className="table one testClassString"
    header={
      <React.Fragment>
        <Link>
          Test
        </Link>
        <Row
          className="header"
          justify="space-between"
        >
          <Search
            className="search"
            dark={false}
            noDelay={true}
            placeholder="Test search"
            search={[MockFunction]}
            text="Test search"
          />
          <Link>
            Test
          </Link>
          <Button />
        </Row>
      </React.Fragment>
    }
  >
    <Box
      className="card table one testClassString"
      justify="flex-start"
    >
      <Flex
        className="card table one testClassString"
        flow="box"
        justify="flex-start"
      >
        <div
          className="flex box flex-start card table one testClassString"
        >
          <header>
            <Link>
              <A
                className="link"
              >
                <a
                  className="link"
                >
                  Test
                </a>
              </A>
            </Link>
            <Row
              className="header"
              justify="space-between"
            >
              <Flex
                className="header"
                flow="row"
                justify="space-between"
              >
                <div
                  className="flex row space-between header"
                >
                  <Search
                    className="search"
                    dark={false}
                    noDelay={true}
                    placeholder="Test search"
                    search={[MockFunction]}
                    text="Test search"
                  >
                    <div
                      className="field search search"
                    >
                      <input
                        className="input"
                        onChange={[Function]}
                        placeholder="Test search"
                        type="text"
                        value="Test search"
                      />
                      <div
                        className="clear"
                        onClick={[Function]}
                      >
                        x
                      </div>
                    </div>
                  </Search>
                  <Link>
                    <A
                      className="link"
                    >
                      <a
                        className="link"
                      >
                        Test
                      </a>
                    </A>
                  </Link>
                  <Button>
                    <Link>
                      <A
                        className="link"
                      >
                        <a
                          className="link"
                        >
                          <button
                            className="button undefined"
                            type="button"
                          />
                        </a>
                      </A>
                    </Link>
                  </Button>
                </div>
              </Flex>
            </Row>
          </header>
          <main
            className="full"
          >
            &lt;p&gt;This is a test&lt;/p&gt;
          </main>
        </div>
      </Flex>
    </Box>
  </Card>
</Table>
`;

exports[`Table should render correctly 2`] = `
<Link
  className="tableRow testClassString"
  to="/"
>
  <Row
    justify="space-between"
  >
    <Row
      className="fields"
      full={true}
      justify="flex-start"
      wrap={false}
    >
      1
      2
      3
    </Row>
    <Box
      auto={true}
      className="arrowCol"
    >
      <Icon
        className="arrow"
        icon="arrow"
        key="arrow"
      />
    </Box>
    <Box
      auto={true}
    >
      <Link
        className="remove"
        onClick={[MockFunction]}
      >
        <Box>
          ×
        </Box>
      </Link>
    </Box>
  </Row>
</Link>
`;

exports[`Table should render correctly 3`] = `
<NoResults>
  <Box
    className="noResults"
  >
    <Flex
      className="noResults"
      flow="box"
    >
      <div
        className="flex box undefined noResults"
      >
        <Icon
          icon="alien_ship"
        >
          <Imported>
            <ReactImportedComponent
              loadable={
                Object {
                  "done": true,
                  "importFunction": [Function],
                  "load": [Function],
                  "mark": Array [],
                  "ok": true,
                  "payload": Object {
                    "default": Object {
                      "process": [Function],
                    },
                    "process": [Function],
                  },
                  "promise": Promise {},
                  "reset": [Function],
                }
              }
              render={[Function]}
            >
              <UnconnectedReactImportedComponent
                async={false}
                exportPicker={[Function]}
                loadable={
                  Object {
                    "done": true,
                    "importFunction": [Function],
                    "load": [Function],
                    "mark": Array [],
                    "ok": true,
                    "payload": Object {
                      "default": Object {
                        "process": [Function],
                      },
                      "process": [Function],
                    },
                    "promise": Promise {},
                    "reset": [Function],
                  }
                }
                render={[Function]}
                streamId={0}
              >
                <img
                  src={
                    Object {
                      "process": [Function],
                    }
                  }
                />
              </UnconnectedReactImportedComponent>
            </ReactImportedComponent>
          </Imported>
        </Icon>
        <H5>
          <h5
            className="h5"
          >
            Nothing found.
          </h5>
        </H5>
      </div>
    </Flex>
  </Box>
</NoResults>
`;

exports[`Table should render default to 1`] = `
<div
  className="tableRow testClassString"
>
  <Row
    justify="space-between"
  >
    <Row
      className="fields"
      full={true}
      justify="flex-start"
      wrap={false}
    >
      1
      2
      3
    </Row>
    <Box
      auto={true}
    >
      <Link
        className="remove"
        onClick={[MockFunction]}
      >
        <Box>
          ×
        </Box>
      </Link>
    </Box>
  </Row>
</div>
`;

exports[`Table should render elements for rows 1`] = `
<Link
  className="tableRow testClassString"
  to="/"
>
  <Row
    justify="space-between"
  >
    <Row
      className="fields"
      full={true}
      justify="flex-start"
      wrap={false}
    >
      <p
        key="1"
      >
        1
      </p>
    </Row>
    <Box
      auto={true}
      className="arrowCol"
    >
      <Icon
        className="arrow"
        icon="arrow"
        key="arrow"
      />
    </Box>
    <Box
      auto={true}
    >
      <Link
        className="remove"
        onClick={[MockFunction]}
      >
        <Box>
          ×
        </Box>
      </Link>
    </Box>
  </Row>
</Link>
`;
